### ISSUE TITLE
Triage (CLI) configuration and pluggability

### ISSUE THREAD

**Issue by nanounanue** [dssg/triage#450]
*Note*: I realize that this issue  should be divided in several subissues. Nevertheless I am creating it because I wanted to have the original @jesteria post in one place accessible to all)

(All the following is from the @jesteria's  slack  post):

***My project***

```
<root>/
└── experiment.py
```

**CLI**

`$ triage [-s|--setup PATH] ...`

  * CLI looks for and optionally loads Python module experiment in CWD
    i.e. argument --setup defaults to `os.path.join(os.curdir, 'experiment.py')` -- ...IFF this file exists
  * If setup desired, but at different path, can specify `--setup`, or envvar `TRIAGE_SETUP`

**Setup module**

Probably should just be a general entrypoint, allowed to execute arbitrary code, (rather than expected to merely define settings, like Django's settings.py).

  * User gets (optionally) a Pythonic entrypoint, e.g. for setup of SQLAlchemy listeners (yay!)
  * Triage gets (optionally) functional specification of its configuration

E.g., `experiment.py:`

```{python}
from sqlalchemy.event import listens_for
from sqlalchemy.pool import Pool

from triage import config

@listens_for(Pool, "connect")
def my_on_connect(dbapi_con, connection_record):
	print("New DBAPI connection:", dbapi_con)

config.update(
	# no way I'm going to set this here -- triage.config respects envvars!
	# database_url=...,
	...
)

# I want an easy, bulk way of updating keys, without overwriting entire dictionaries (as __setitem__ and update would do)
# (Note, we could probably achieve the same with `update`, so this interface really isn't necessary, depending on how people like it.)
config.set(
	temporal_config__feature_start_time='1905-01-01',
	...
)

if os.getenv('PUDDING') == 'plentiful':
	config['eat'] = 'it-all'
```

**Triage config**

  * provides dict interface
  * but backed by reasonable defaults: user setttings > envvar settings > library defaults
  * Perhaps CLI imports triage.config, after loading experiment.py, and uses config to instantiate lower-   level triage classes

We get:

  * Pythonic pluggability in CLI
  * the centralization of configuration
  * a place to insert envvars
  * lower-level classes unaffected by these concerns (of envvars and configuration, etc.)

**Comment by nanounanue** [comment]
@jesteria continues: 

oh and `yaml` is really nice ... perhaps `triage.config` should also have a method for loading from yaml, (not that that's hard, but whatever makes it easiest) -- and/or `--setup` could support the specification of a directory of yaml files, as we've discussed, for centralized configuration by yaml, (without the python module, if not needed)

**Comment by ecsalomon** [comment]
This is slightly off-topic, but I wanted to not somewhere that 

```
@listens_for(Pool, "connect")
def my_on_connect(dbapi_con, connection_record):
	print("New DBAPI connection:", dbapi_con)
```

does not catch all of the Triage database events. When predictions are made, neither the `delete` on old predictions nor the `copy` of new predictions is captured by this event type. I've experimented with trying to capture these events, but haven't yet found the right class/event combination for either of them.

**Comment by nanounanue** [comment]
> This is slightly off-topic, but I wanted to note somewhere that
> 
> ```
> @listens_for(Pool, "connect")
> def my_on_connect(dbapi_con, connection_record):
> 	print("New DBAPI connection:", dbapi_con)
> ```
> 
> does not catch all of the Triage database events. When predictions are made, neither the `delete` on old predictions nor the `copy` of new predictions is captured by this event type. I've experimented with trying to capture these events, but haven't yet found the right class/event combination for either of them.

**Comment by nanounanue** [comment]
(My mistake)

**Comment by nanounanue** [comment]
> The delete we can probably accomplish by rewriting that query in triage (although it should also be easier to accomplish in SQLAlchemy events, given that we aren't checking out a raw connection).
> 
> The COPY is more worrying, as we really need the raw psycopg2 connection in order to do the COPY. That's what I'm looking at primarily.

Yup. I didn't think abot the `COPY`. It should  be ways of running it from `sqlalchemy`. Maybe we can extract the `raw_connection` 

See

https://stackoverflow.com/a/22074610/754176

But I din't try it

**Comment by ecsalomon** [comment]
Just a quick update: after the 408 branch was merged, i was able to run the whole thing with run.py. Only the Pool/connect listener was ever accessed. I don't know whether some difference between the 408 branch (what i was running before) and master caused the issue I experienced, but I can no longer replicate it.
